### PR TITLE
Correctly find enum members in case labels

### LINKED ISSUE #53269: Binding broken with enum members inside switch case labels
**Version Used**: 16.10 Preview 2

**Steps to Reproduce**:

1. Create a new C# project.
2. Paste this in to a method somewhere
```c#
FileAttributes f = FileAttributes.Archive;

switch (f)
{
    case FileAttributes.Archive:
        break;
}

```
3. Click on "Archive" in the case label.

**Expected Behavior**: We find references to "Archive".
**Actual Behavior**: We oddly highlight the entire line. Searches aren't found.

Note this is **not** unique to just highlight refs: quick info shows binding to the label, and find references invoked at that location no longer works.

The reason this is happening is because case statements are now declaring symbols, and SymbolFinder.FindSymbolAtPosition (which a lot of these features call into) prefers declaring symbols in that case. Why is this a declaring symbol? Because of the code here:

https://github.com/dotnet/roslyn/blob/25d199474486ec26617ade5a485f57fc118e64c5/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Services/SemanticFacts/CSharpSemanticFacts.cs#L54-L80

The algorithm is to simply walk up the tree, and start finding whatever declared symbols we can find. The expectation however is the "location" of a symbol never contains an expression like this, but in this label case it does.

### PR DESCRIPTION
**Pull request by jasonmalinowski**
When we changed the logic for finding declared symbols in ca2505a08c8974dde3243bd9fdb6a159c3d83cd7, we now would see the case label as declaring the symbol which we don't want. As a conservative choice, we'll do the expanded logic only for operators where it was originally intended.

Fixes #53269

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
> jasonmalinowski force-pushed 

:( :( :(

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Services/SemanticFacts/CSharpSemanticFacts.cs [hunk 1]
```diff
@@ -60,10 +60,21 @@ public ISymbol GetDeclaredSymbol(SemanticModel semanticModel, SyntaxToken token,
...
+                        if (symbol.Locations.Any(loc => loc.SourceTree == location.SourceTree && loc.SourceSpan.Contains(location.SourceSpan)))
+                            return symbol;
+                    }
+                    else
+                    {
+                        // For any other symbols, we only care if the name directly matches the span of the token
+                        if (symbol.Locations.Contains(location))
+                        {
+                            return symbol;
+                        }
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Services/SemanticFacts/CSharpSemanticFacts.cs:76]:
```suggestion
                            return symbol;
```

to match the formatting you have above.

**Review by CyrusNajmabadi — APPROVED** [review]

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Services/SemanticFacts/CSharpSemanticFacts.cs (hunk 1 above)
Comment (jasonmalinowski) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Services/SemanticFacts/CSharpSemanticFacts.cs]: Fixed.

**Comment by jasonmalinowski** [comment]
Tagging @jinujoseph for approval for 16.11. We'll put it here first and decide if we want to backport later.

**Pull request merged**